cmd/ido2icingadb/misc.go

```diff
@@ -0,0 +1,354 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/sha1"
+	"encoding/binary"
+	"github.com/google/uuid"
+	"github.com/icinga/icingadb/pkg/icingadb/objectpacker"
+	icingadbTypes "github.com/icinga/icingadb/pkg/types"
+	"github.com/jmoiron/sqlx"
+	"github.com/pkg/errors"
+	"github.com/vbauerster/mpb/v6"
+	"github.com/vbauerster/mpb/v6/decor"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"reflect"
+	"sync"
+	"time"
+)
+
+type ProgressRow struct {
+	Id   uint64
+	Name string
+}
+
+type convertedId struct {
+	ido uint64
+	idb []byte
+}
+
+type barIncrementer struct {
+	bar   *mpb.Bar
+	start time.Time
+}
+
+func (bi *barIncrementer) inc(i int) {
+	prev := bi.start
+	now := time.Now()
+	bi.start = now
+
+	bi.bar.IncrBy(i)
+	bi.bar.DecoratorEwmaUpdate(now.Sub(prev))
+}
+
+const bulk = 10000
+
+var log = func() *zap.SugaredLogger {
+	logger, err := zap.NewDevelopmentConfig().Build()
+	if err != nil {
+		panic(err)
+	}
+
+	return logger.Sugar()
+}()
+
+var objectTypes = map[uint8]string{1: "host", 2: "service"}
+
+// mkDeterministicUuid returns a formally random UUID (v4) as follows: 11111122-3300-4455-4455-555555555555
+//
+// 0: zeroed
+// 1: "IDO" (where the data identified by the new UUID is from)
+// 2: the history table the new UUID is for, e.g. "s" for state_history
+// 3: "h" (for "history")
+// 4: the new UUID's formal version (unused bits zeroed)
+// 5: the ID of the row the new UUID is for in the IDO (big endian)
+func mkDeterministicUuid(table byte, rowId uint64) icingadbTypes.UUID {
+	uid := uuidTemplate
+	uid[3] = table
+
+	buf := &bytes.Buffer{}
+	if err := binary.Write(buf, binary.BigEndian, rowId); err != nil {
+		panic(err)
+	}
+
+	bEId := buf.Bytes()
+	uid[7] = bEId[0]
+	copy(uid[9:], bEId[1:])
+
+	return icingadbTypes.UUID{UUID: uid}
+}
+
+// uuidTemplate is for mkDeterministicUuid.
+var uuidTemplate = func() uuid.UUID {
+	buf := &bytes.Buffer{}
+	buf.Write(uuid.Nil[:])
+
+	uid, err := uuid.NewRandomFromReader(buf)
+	if err != nil {
+		panic(err)
+	}
+
+	copy(uid[:], "IDO h")
+	return uid
+}()
+
+// randomUuid generates a new UUIDv4.
+func randomUuid() icingadbTypes.UUID {
+	var rander *bufio.Reader
+
+	massRanders.Lock()
+	for r := range massRanders.pool {
+		rander = r
+		delete(massRanders.pool, r)
+		break
+	}
+	massRanders.Unlock()
+
+	if rander == nil {
+		rander = bufio.NewReader(rand.Reader)
+	}
+
+	id, err := uuid.NewRandomFromReader(rander)
+	if err != nil {
+		log.Fatalf("%+v", errors.Wrap(err, "can't generate random UUID"))
+	}
+
+	massRanders.Lock()
+	massRanders.pool[rander] = struct{}{}
+	massRanders.Unlock()
+
+	return icingadbTypes.UUID{UUID: id}
+}
```

Comment:
System calls are **always** bottle necks.

Reply:
Okay, so Go indeed performs one `getrandom(2)` syscall per call to `crypto/rand.Reader.Read()`, interesting, that's not what I would have expected. But still, one has to find the right balance between complexity and performance (I mean we aren't implementing Icinga DB as a kernel module after all, even though that would greatly reduce the number of syscalls). But let's await our HA/History discussion first.

Reply:
"the best request is no request", the best malloc() is no malloc() (Icinga/icinga2#8152) and the best syscall is... (guess once). I'd prefer to have just one variable initialized with a one-liner, but things turned out to be a litte more complex.

TBH I'm really wondering on statements like yours. Or on @pdolinic ones. Asked him on TLS about RSA vs. ECC and he: ECC is just used for IOT. Excuse me: sysadmins have a not less secure, but more efficient tech. available, but just don't use it? As my ex-classmate from Sparda EDV said (on server-side-only PHP form validation): [B2B] internet is cheap, storage is cheap, but CPU is expensive. Also CPU consumes energy. Choose for yourself whether you want to save that energy because of the current prices or because of Greta (or both).

Reply:
I think that this might just be premature optimization. Just because something can be optimized slightly, it's not necessarily a good idea to do so.